src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+    this.read_buffer.clear();
+
+    this.command = DSP_NO_COMMAND;
+    this.command_size = 0;
+
+    this.dummy_speaker_enabled = false;
+    this.test_register = 0;
+
+    this.dsp_highspeed = false;
+    this.dsp_stereo = false;
+    this.dsp_16bit = false;
+    this.dsp_signed = false;
+
+    this.dac_buffer.clear();
+    this.dac_rate_ratio = 2;
+
+    this.dma_sample_count = 0;
+    this.dma_bytes_count = 0;
+    this.dma_bytes_left = 0;
+    this.dma_bytes_block = 0;
+    this.dma_irq = 0;
+    this.dma_channel = 0;
+    this.dma_autoinit = false;
+    this.dma_buffer_uint8.fill(0);
+
+    this.e2_value = 0xAA;
+    this.e2_count = 0;
+
+    this.sampling_rate = 22050;
+    this.bytes_per_sample = 1;
+
+    this.lower_irq(SB_IRQ_8BIT);
+    this.irq_triggered.fill(0);
+
+    this.asp_registers.fill(0);
+    this.asp_registers[5] = 0x01;
+    this.asp_registers[9] = 0xF8;
+}
+
+SB16.prototype.get_state = function()
+{
+    var state = [];
+
+    // state[ 0] = this.write_buffer              ;
+    // state[ 1] = this.read_buffer               ;
+    state[ 2] = this.read_buffer_lastvalue     ;
+
+    state[ 3] = this.command                   ;
+    state[ 4] = this.command_size              ;
+
+    state[ 5] = this.mixer_current_address     ;
+    state[ 6] = this.mixer_unhandled_registers ;
+
+    state[ 7] = this.dummy_speaker_enabled     ;
+    state[ 8] = this.test_register             ;
+
+    state[ 9] = this.dsp_highspeed             ;
+    state[10] = this.dsp_stereo                ;
+    state[11] = this.dsp_16bit                 ;
+    state[12] = this.dsp_signed                ;
+
+    // state[13] = this.dac_buffer                ;
+    state[14] = this.dac_rate_ratio            ;
+
+    state[15] = this.dma_sample_count          ;
+    state[16] = this.dma_bytes_count           ;
+    state[17] = this.dma_bytes_left            ;
+    state[18] = this.dma_bytes_block           ;
+    state[19] = this.dma_irq                   ;
+    state[20] = this.dma_channel               ;
+    state[21] = this.dma_channel_8bit          ;
+    state[22] = this.dma_channel_16bit         ;
+    state[23] = this.dma_autoinit              ;
+    state[24] = this.dma_buffer_uint8          ;
+    state[25] = this.sampling_rate             ;
+    state[26] = this.bytes_per_sample          ;
+
+    state[27] = this.e2_value                  ;
+    state[28] = this.e2_count                  ;
+
+    state[29] = this.asp_registers             ;
+
+    // state[30] = this.mpu_read_buffer           ;
+    state[31] = this.mpu_read_buffer_last_value;
+
+    state[32] = this.irq                       ;
+    state[33] = this.irq_triggered             ;
+    state[34] = this.audio_samplerate          ;
+
+    return state;
+};
+
+SB16.prototype.set_state = function(state)
+{
+    // this.write_buffer               = state[ 0];
+    // this.read_buffer                = state[ 1];
+    this.read_buffer_lastvalue      = state[ 2];
```

Comment:
Follow the general style of code in this project: Only a single space before `=`.